diff --git a/drivers/staging/qcacld-3.0/Makefile b/drivers/staging/qcacld-3.0/Makefile
index fb78b03460b6..75d4871a1c50 100644
--- a/drivers/staging/qcacld-3.0/Makefile
+++ b/drivers/staging/qcacld-3.0/Makefile
@@ -11,6 +11,7 @@ KERNEL_SRC ?= /lib/modules/$(shell uname -r)/build
 # module source directory.
 M ?= $(shell pwd)
 
+KBUILD_CFLAGS   += -fno-lto
+KBUILD_AFLAGS   += -fno-lto
 # WLAN_ROOT must contain an absolute path (i.e. not a relative path)
 KBUILD_OPTIONS := WLAN_ROOT=$(shell cd $(KERNEL_SRC); readlink -e $(M))
 KBUILD_OPTIONS += MODNAME?=wlan
