diff --git a/lib/Makefile b/lib/Makefile
index d78b53461f86..5f126bb7af41 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -7,7 +7,6 @@ ifdef CONFIG_FUNCTION_TRACER
 ORIG_CFLAGS := $(KBUILD_CFLAGS)
 KBUILD_CFLAGS = $(subst $(CC_FLAGS_FTRACE),,$(ORIG_CFLAGS))
 endif
-
 # These files are disabled because they produce lots of non-interesting and/or
 # flaky coverage that is not a function of syscall inputs. For example,
 # rbtree can be global and individual rotations don't correlate with inputs.
@@ -23,11 +22,11 @@ KASAN_SANITIZE_string.o := n
 
 ifdef CONFIG_FUNCTION_TRACER
 CFLAGS_REMOVE_string.o = -pg
+CFLAGS_win_minmax.o += -fPIC
 endif
 
 CFLAGS_string.o := $(call cc-option, -fno-stack-protector)
 endif
-
 lib-y := ctype.o string.o vsprintf.o cmdline.o \
 	 rbtree.o radix-tree.o dump_stack.o timerqueue.o\
 	 idr.o int_sqrt.o extable.o \
