From a21b187efb4eadffb30a35839530bd577838352b Mon Sep 17 00:00:00 2001
From: TheSillyOk <priv.ld@proton.me>
Date: Sat, 8 Feb 2025 17:32:35 -0300
Subject: [PATCH] Fix rejects

---
 fs/proc/base.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/fs/proc/base.c b/fs/proc/base.c
index 850d80b530e7..b4c9068081f3 100644
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -2205,11 +2205,18 @@ static int map_files_get_link(struct dentry *dentry, struct path *path)
 	rc = -ENOENT;
 	down_read(&mm->mmap_sem);
 	vma = find_exact_vma(mm, vm_start, vm_end);
-	if (vma && vma->vm_file) {
-		*path = vma->vm_file->f_path;
-		path_get(path);
-		rc = 0;
-	}
+	if (vma) {
+        if (vma->vm_file) {
+            if (strstr(vma->vm_file->f_path.dentry->d_name.name, "lineage")) { 
+				rc = kern_path("/system/framework/framework-res.apk", LOOKUP_FOLLOW, path);
+			} else {
+				*path = vma->vm_file->f_path;
+				path_get(path);
+				rc = 0;
+            }
+        }
+    }
+
 	up_read(&mm->mmap_sem);
 
 out_mmput: